File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/booking/BookingRows.kt
package lk.mageride.passenger.booking

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.DirectionsWalk
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import lk.mageride.passenger.R
import lk.mageride.passenger.home.VehicleLabels
import lk.mageride.passenger.ride.PaymentRails
import lk.mageride.passenger.ui.MoneyFormat
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.RideVehicleType
import lk.mageride.shared.data.models.VehicleType
import lk.mageride.shared.data.models.fare.PaymentMethod
import lk.mageride.shared.data.models.transit.TransitOptionKind

// SCR-PA-009's two lists, split out of `RideBookingScreen.kt` so neither file is a wall.
//
// The screen is the layout — map, summary, toggles, CTA — and this is what goes between: a GTFS
// section that degrades to one muted row (AL-55), and a tier row that can show a price and nothing
// else (AL-19). Both are the fences, drawn.

/**
 * The GTFS half.
 *
 * **AL-55's degradation is a muted row, not an error.** D2' §SCR-PA-009: when transit-svc has no
 * feed for the corridor or is unreachable, the section becomes *"Bus route info coming soon for
 * this area"* and everything else on the screen keeps working. Nothing blocks on GTFS coverage.
 */
@Composable
internal fun PublicSection(state: RideBookingState, onSelect: (PublicRouteRow) -> Unit) {
    SectionLabel(text = stringResource(R.string.booking_public_section))

    when {
        state.routesLoading -> Loading(stringResource(R.string.booking_loading_routes))

        state.publicUnavailable -> MutedRow(stringResource(R.string.booking_no_gtfs))

        state.routes.isEmpty() -> MutedRow(stringResource(R.string.booking_no_routes))

        else -> state.routes.forEach { route ->
            PublicRouteCard(
                route = route,
                selected = (state.selection as? BookingSelection.Public)?.route == route,
                onSelect = { onSelect(route) },
            )
        }
    }

    // "⓵ Walk 250 m to Pamankada halt — blue route on map", shown only when off-route.
    state.walkHalt?.let { hint ->
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Icon(
                imageVector = Icons.AutoMirrored.Filled.DirectionsWalk,
                contentDescription = null,
                modifier = Modifier.size(ControlTokens.ChipIcon),
                tint = MaterialTheme.colorScheme.primary,
            )
            Text(
                text = stringResource(R.string.booking_walk_hint, hint.metres, hint.haltName),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

/** `🚌 138 · Pettah → Maharagama | via High Level Rd | Direct | PUBLIC`. */
@Composable
internal fun PublicRouteCard(route: PublicRouteRow, selected: Boolean, onSelect: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onSelect)
            .padding(vertical = MageRideTheme.spacing.xs),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        RadioButton(selected = selected, onClick = onSelect)
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = listOfNotNull(
                    route.routeShortName.takeIf { it.isNotBlank() },
                    route.headsign,
                ).joinToString(ROUTE_SEPARATOR),
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
            route.description?.takeIf { it.isNotBlank() }?.let {
                Text(
                    text = it,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            if (route.transfers > 0) {
                Text(
                    text = stringResource(R.string.booking_transfers, route.transfers),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
        Tag(
            text = stringResource(
                if (route.kind ==
                    TransitOptionKind.DIRECT
                ) {
                    R.string.booking_tag_direct
                } else {
                    R.string.booking_tag_transit
                },
            ),
        )
        Tag(text = stringResource(R.string.booking_tag_public))
    }
}

/**
 * One Mode C tier — **type, "Standby · upfront fare", and a price. Nothing else.**
 *
 * AL-19 / BR-23.3: no minutes-away, no distance to driver, because no driver has been matched. The
 * card could not show one if it wanted to — [TierQuote] has no such field.
 */
@Composable
internal fun TierRow(quote: TierQuote, selected: Boolean, onSelect: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onSelect)
            .padding(vertical = MageRideTheme.spacing.xs),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        RadioButton(selected = selected, onClick = onSelect)
        Icon(
            imageVector = VehicleLabels.icon(quote.vehicleType.asVehicleType()),
            contentDescription = null,
            modifier = Modifier.size(ControlTokens.RowIcon),
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = stringResource(VehicleLabels.name(quote.vehicleType.asVehicleType())),
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )
            Text(
                text = stringResource(R.string.booking_standby),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Text(
            text = MoneyFormat.rupees(quote.amountMinor),
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

/** The wireframe's `Payment: Cash ▾` chip. C080's SCR-PA-016 is the full picker. */
@Composable
internal fun PaymentChip(method: PaymentMethod, onChange: (PaymentMethod) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        // The three rails that survive AL-57/AL-59, and no surcharge on any of them. COD is
        // absent because it is package-only (AL-22, US-20.8) — SCR-PA-012 has its own row.
        PaymentRails.RIDE.forEach { option ->
            ToggleChip(
                label = stringResource(PaymentRails.label(option)),
                selected = option == method,
                onClick = { onChange(option) },
            )
        }
    }
}

/** `Direct` / `Transit` / `PUBLIC`. */
@Composable
internal fun Tag(text: String) {
    Text(
        text = text,
        modifier = Modifier
            .background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(MageRideTheme.radius.sm))
            .padding(horizontal = MageRideTheme.spacing.xxs, vertical = MageRideTheme.spacing.xxs),
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}

@Composable
internal fun Loading(text: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        CircularProgressIndicator(modifier = Modifier.size(ControlTokens.ChipIcon))
        Text(
            text = text,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
internal fun MutedRow(text: String) {
    Text(
        text = text,
        modifier = Modifier.padding(vertical = MageRideTheme.spacing.xs),
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}

/**
 * The bookable subset back to the full AL-09 type, so the tier row can reuse C078's icon and name
 * tables rather than growing a second copy of §0.2's vehicle list.
 */
internal fun RideVehicleType.asVehicleType(): VehicleType = when (this) {
    RideVehicleType.MOTORBIKE -> VehicleType.MOTORBIKE
    RideVehicleType.THREE_WHEELER -> VehicleType.THREE_WHEELER
    RideVehicleType.FLEX -> VehicleType.FLEX
    RideVehicleType.SEDAN -> VehicleType.SEDAN
    RideVehicleType.MINI_VAN -> VehicleType.MINI_VAN
    RideVehicleType.VAN -> VehicleType.VAN
    RideVehicleType.TRUCK -> VehicleType.TRUCK
    RideVehicleType.MINI_TRUCK -> VehicleType.MINI_TRUCK
}

/** The wireframe's `·` between a route number and its headsign. Punctuation, not copy. */
private const val ROUTE_SEPARATOR = " · "

File: apps/passenger-ios/PassengerApp/Booking/BookingRows.swift
import MageRideShared
import SwiftUI

// SCR-PI-009's two lists, split out of `RideBookingScreen.swift` so neither file is a wall.
//
// The screen is the layout — map, summary, pickers, CTA — and this is what goes between: a GTFS
// section that degrades to one muted row (AL-55), and a tier row that can show a price and nothing
// else (AL-19). Both are the fences, drawn.

/// The GTFS half.
///
/// **AL-55's degradation is a muted row, not an error.** D2' §SCR-PA-009: when transit-svc has no
/// feed for the corridor or is unreachable, the section becomes *"Bus route info coming soon for
/// this area"* and everything else on the screen keeps working. Nothing blocks on GTFS coverage.
struct PublicSection: View {

    let state: RideBookingState
    let onSelect: (PublicRouteRow) -> Void

    var body: some View {
        SectionLabel(key: "booking_public_section")

        if state.routesLoading {
            LoadingRow(messageKey: "booking_loading_routes")
        } else if state.publicUnavailable {
            MutedRow(messageKey: "booking_no_gtfs")
        } else if state.routes.isEmpty {
            MutedRow(messageKey: "booking_no_routes")
        } else {
            ForEach(state.routes) { route in
                PublicRouteCard(
                    route: route,
                    isSelected: state.selection == .publicRoute(route),
                    onSelect: { onSelect(route) }
                )
            }
        }

        // "⓵ Walk 250 m to Pamankada halt — blue route on map", shown only when off-route.
        if let hint = state.walkHalt {
            FormattedBanner(
                messageKey: "booking_walk_hint",
                arguments: [hint.metres, hint.haltName],
                symbolName: "figure.walk"
            )
        }
    }
}

/// `🚌 138 · Pettah → Maharagama | via High Level Rd | Direct | PUBLIC`.
struct PublicRouteCard: View {

    let route: PublicRouteRow
    let isSelected: Bool
    let onSelect: () -> Void

    var body: some View {
        TierCard(
            // A GTFS option is a bus or a train, and the feed does not say which; §0.2's bus green
            // is the section's own colour and the badge under it says PUBLIC either way.
            symbolName: VehicleToken.bus.symbolName,
            symbolColor: VehicleToken.bus.color,
            title: title,
            subtitle: subtitle,
            isSelected: isSelected,
            action: onSelect
        ) {
            VStack(alignment: .trailing, spacing: MageRideSpacing.xxs) {
                StatusPill(
                    titleKey: route.isDirect ? "booking_tag_direct" : "booking_tag_transit",
                    tone: route.isDirect ? .ok : .muted
                )
                SolidBadge(text: "booking_tag_public".localised, color: VehicleToken.bus.color)
            }
        }
    }

    private var title: String {
        [route.routeShortName.isEmpty ? nil : route.routeShortName, route.headsign]
            .compactMap { $0 }
            .joined(separator: MageRideSymbols.separator)
    }

    /// The feed's own description, and the transfer count under it when there is one. The
    /// wireframe's *"every ~10 min"* is **not** drawn: no contract carries a headway (C079's gap).
    private var subtitle: String? {
        let parts = [
            route.routeDescription?.isEmpty == false ? route.routeDescription : nil,
            route.transfers > 0 ? "booking_transfers".localisedFormat(route.transfers) : nil,
        ]
        let joined = parts.compactMap { $0 }.joined(separator: MageRideSymbols.separator)
        return joined.isEmpty ? nil : joined
    }
}

/// One Mode C tier — **type, "Standby · upfront fare", and a price. Nothing else.**
///
/// AL-19 / BR-23.3: no minutes-away, no distance to driver, because no driver has been matched. The
/// card could not show one if it wanted to — ``TierQuote`` has no such field.
struct TierRow: View {

    let quote: TierQuote
    let isSelected: Bool
    let onSelect: () -> Void

    var body: some View {
        let token = VehicleToken.forType(quote.vehicleType.toVehicleType()) ?? .privateHire

        TierCard(
            symbolName: token.symbolName,
            symbolColor: token.color,
            title: token.nameKey.localised,
            subtitle: "booking_standby".localised,
            isSelected: isSelected,
            action: onSelect
        ) {
            Text(MoneyFormat.rupees(quote.amountMinor))
                .mageFont(.subtitle)
                .foregroundStyle(MageRideColor.onSurface)
        }
    }
}

/// The wireframe's `Payment: Cash ▾` chip row. C098's SCR-PI-016 is the full picker.
///
/// The three rails that survive AL-57/AL-59, and no surcharge on any of them. COD is absent because
/// it is package-only (AL-22, US-20.8) — SCR-PI-012 has its own row.
struct PaymentChipRow: View {

    let method: PaymentMethod
    var rails: [PaymentMethod] = PaymentRails.ride
    let onChange: (PaymentMethod) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: MageRideSpacing.xs) {
                ForEach(rails, id: \.wire) { rail in
                    PlaceChip(
                        label: PaymentRails.labelKey(rail).localised,
                        symbolName: rail == method ? "checkmark" : "creditcard"
                    ) {
                        onChange(rail)
                    }
                    .opacity(rail == method ? 1 : 0.6)
                }
            }
            .padding(.vertical, 1)
        }
    }
}

/// A read in flight — the wireframe's shimmer, as the platform's own spinner.
struct LoadingRow: View {

    let messageKey: String

    var body: some View {
        HStack(spacing: MageRideSpacing.xs) {
            ProgressView()
            Text(key: messageKey)
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
            Spacer(minLength: 0)
        }
        .frame(minHeight: MageRideControl.minimumTapTarget)
    }
}

/// A section that has nothing in it, and says which kind of nothing.
struct MutedRow: View {

    let messageKey: String

    var body: some View {
        Text(key: messageKey)
            .mageFont(.bodySmall)
            .foregroundStyle(MageRideColor.onSurfaceVariant)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.vertical, MageRideSpacing.xs)
    }
}

/// The wireframe's `card fill` summary — `● Galle Face` over `◉ Nugegoda ✎`, each end named.
///
/// **Both ends carry a label, and that is a defect fix rather than decoration** (Δ handset report,
/// found on the Android twin and ported here). A pickup with no street address prints
/// *"Current location"* — which ``LastKnownFix`` is right to do, since the fix is a coordinate
/// nobody has geocoded — and it is ALWAYS what the top row shows, because nothing in the app writes
/// an address onto a pickup. Two rows of the same type, one above the other, therefore read as a
/// caption over the place the passenger was travelling to. The `cdot`s were already here; the
/// labels are the rest of the answer.
///
/// **The ✎ moved to the destination row**, because `onEdit` sets ``CaptureTarget/bookingDropoff``
/// — it opens SCR-PI-008 to change where the journey ENDS. The wireframe draws it beside the
/// pickup, which is the cell describing a control the flow behind it does not have.
struct JourneySummaryCard: View {

    let pickup: Place?

    /// Whether ``pickup`` is a place somebody PICKED rather than the passenger's own fix — see
    /// ``BookingDraftState/pickupIsChosen``. It decides whether a pickup with no address reads as
    /// *"Current location"* or as the coordinates it was pinned at.
    let pickupIsChosen: Bool

    let dropoff: Place?
    let onEdit: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
            SectionLabel(key: "booking_pickup_label")

            HStack(spacing: MageRideSpacing.xs) {
                Circle()
                    .fill(MageRideColor.success)
                    .frame(width: MageRideControl.routeDot, height: MageRideControl.routeDot)
                Text(pickupText)
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.onSurface)
                    .lineLimit(1)
                Spacer(minLength: 0)
            }

            SectionLabel(key: "booking_destination_label")

            HStack(spacing: MageRideSpacing.xs) {
                Circle()
                    .fill(MageRideColor.error)
                    .frame(width: MageRideControl.routeDot, height: MageRideControl.routeDot)
                Text(dropoff.map(PlaceLabel.of) ?? "booking_no_destination".localised)
                    .mageFont(.bodySmall)
                    .foregroundStyle(dropoff == nil ? MageRideColor.onSurfaceVariant : MageRideColor.onSurface)
                    .lineLimit(1)
                Spacer(minLength: MageRideSpacing.xs)
                Button(action: onEdit) {
                    Image(systemName: "pencil")
                        .font(.system(size: MageRideControl.chipIcon))
                        .foregroundStyle(MageRideColor.primary)
                }
                .accessibilityLabel(Text(key: "booking_edit_route"))
            }
        }
        .padding(MageRideSpacing.sm)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    /// A pickup somebody PICKED is named, or shown as the coordinates it was pinned at — only the
    /// passenger's own fix is *"Current location"*.
    ///
    /// A proxy rider's pin has no address and is not where the booker is standing, so printing the
    /// fix's label over it was a lie the booker could not correct.
    private var pickupText: String {
        guard let pickup else { return "search_current_location".localised }
        if pickupIsChosen { return PlaceLabel.of(pickup) }
        return pickup.address ?? "search_current_location".localised
    }
}

/// The Search / Map / Paste link / Request control, shared by SCR-PI-010b and SCR-PI-012.
///
/// One view rather than three copies, because the *set of methods* is a rule and not a layout: the
/// fence says a package **pickup** offers three and a **drop-off** offers four, and expressing that
/// as `methods:` at two call sites is what makes it checkable. See ``PackageEnd`` for why the two
/// ends differ.
///
/// A segmented `Picker`, which is the cell's own `seg` — `UISegmentedControl` in the wireframe's CSS.
struct LocationMethodPicker: View {

    let methods: [PickupMethod]
    let selection: PickupMethod
    let onSelect: (PickupMethod) -> Void

    var body: some View {
        Picker("", selection: Binding(get: { selection }, set: { onSelect($0) })) {
            ForEach(methods) { method in
                Text(key: method.labelKey).tag(method)
            }
        }
        .pickerStyle(.segmented)
        .labelsHidden()
    }
}

/// What a captured place reads as: its **name**, or the coordinates it was pinned at.
///
/// A place with no address is not a place with no answer — a pin dropped on an unnamed lane is
/// exactly what the Map method is for, and *"Pinned at 6.92710, 79.86120"* is a passenger telling a
/// driver where to come. Every screen that shows a captured place goes through here, so
/// SCR-PI-009's summary and SCR-PI-010b's row cannot drift apart.
enum PlaceLabel {

    static func of(_ place: Place) -> String {
        if let address = place.address, !address.isEmpty { return address }
        return "capture_pinned".localisedFormat(place.lat, place.lng)
    }
}

/// What a captured place reads as under its control, or the prompt when there is none yet.
struct CapturedPlaceRow: View {

    let place: Place?
    let emptyKey: String

    var body: some View {
        Text(text)
            .mageFont(.bodySmall)
            .foregroundStyle(place == nil ? MageRideColor.onSurfaceVariant : MageRideColor.onSurface)
            .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var text: String {
        guard let place else { return emptyKey.localised }
        return PlaceLabel.of(place)
    }
}
